examples/md-flexible/src/domainDecomposition/RegularGridDecomposition.cpp

```diff
@@ -275,31 +278,128 @@ void RegularGridDecomposition::reflectParticlesAtBoundaries(AutoPasType &autoPas
   using autopas::utils::Math::isNear;
   std::array<double, _dimensionCount> reflSkinMin{}, reflSkinMax{};
-  auto functorLJ = autopas::LJFunctor<ParticleType, false, true, autopas::FunctorN3Modes::Both, false, false>(
-      _maxReflectiveSkin * 2., particlePropertiesLib);
 
   for (int dimensionIndex = 0; dimensionIndex < _dimensionCount; ++dimensionIndex) {
     // skip if boundary is not reflective
     if (_boundaryType[dimensionIndex] != options::BoundaryTypeOption::reflective) continue;
 
     auto reflect = [&](bool isUpper) {
+      const auto boundaryPosition = isUpper ? reflSkinMax[dimensionIndex] : reflSkinMin[dimensionIndex];
+
       for (auto p = autoPasContainer.getRegionIterator(reflSkinMin, reflSkinMax, autopas::IteratorBehavior::owned);
            p.isValid(); ++p) {
         // Check that particle is within 6th root of 2 * sigma
         const auto position = p->getR();
-        const auto distanceToBoundary = isUpper ? reflSkinMax[dimensionIndex] - position[dimensionIndex]
-                                                : position[dimensionIndex] - reflSkinMin[dimensionIndex];
-        if (distanceToBoundary < sixthRootOfTwo * particlePropertiesLib.getSigma(p->getTypeId()) / 2.) {
-          // Create mirror particle and shift it to other side of reflective boundary
-          ParticleType mirrorParticle;
-          auto position = p->getR();
-          position[dimensionIndex] =
-              isUpper ? reflSkinMax[dimensionIndex] + (reflSkinMax[dimensionIndex] - position[dimensionIndex])
-                      : reflSkinMin[dimensionIndex] - position[dimensionIndex];
-          mirrorParticle.setR(position);
+        const auto distanceToBoundary = std::abs(position[dimensionIndex] - boundaryPosition);
+
+        // For single-site molecules, we discard molecules further than sixthRootOfTwo * sigma, and are left only with
+        // molecules who will experience repulsion from the boundary.
+
+        // For multi-site molecules, we discard molecules with center-of-mass further than sixthRootOfTwo * the largest
+        // sigma of any site of that molecule. Some molecules may still experience attraction in which case their force
+        // is reset to before the force interaction.
+        //
+        // Note, there is a scenario where a molecule has center-of-mass further than sixthRootOfTwo * sigma, but has a
+        // site closer than this distance, with a large enough epsilon, that repulsion would occur. For computational
+        // cost reasons, this scenario is neglected - no repulsion occurs. This *should*, in theory, with an appropriate
+        // step-size and molecular model, not cause any problems.
+
+        // To produce a "mirrored" multi-site molecule that could be used with the multi-site molecule functor would be
+        // very nasty. As such, we reimplement the kernel of the lennard-jones force here. We also use this for
```

Comment:
what exactly makes this nasty?

Reply:
We would require mirrored particle types for all 3-dimensions in the PPL. See the updated version of this comment for more info.